Dropdown array for Select limit

eather009 opened the pull request:
```
@param array $limits This is array of options.
@param array $default Default limit for option selecting. Default is Paginator limit.
@param array $options Options for Select tag attributes like class, id or event

@return string html output
```

In Template, just use this code to get dropdown option

```
$this->Paginator->limitControl();
```

This new function is useful if you need options for how many rows will show in Pagination. You can define $limits as array and also you can set $default selected value. 

<img width="168" alt="show limit options" src="https://cloud.githubusercontent.com/assets/4234756/22939416/ce9d975e-f308-11e6-84a7-8f14c64aa01b.png">

Hope it will help users to code less and not to depend on other plugin. 

Thank you

eather009 committed:
need vendor to get updated

eather009 committed:
Dropdown array for Select limit

@param array $limits This is array of options.
@param array $default Default limit for option selecting. Default is Paginator limit.

In Template, just use this code to get dropdown option
<?= $this->Paginator->showLimit(); ?>

eather009 committed:
Dropdown array for Select limit

@param array $limits This is array of options.
@param array $default Default limit for option selecting. Default is Paginator limit.

In Template, just use this code to get dropdown option
<?= $this->Paginator->showLimit(); ?>

eather009 committed:
Revert "need vendor to get updated"

This reverts commit 16cb7d55f1d92fb9bbc27f1d11340b0f3703b446.

eather009 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,37 @@ public function implementedEvents()
+     *
+     * @param array $limits This is array of options.
+     * @param array $default Default limit for option selecting. Default is Paginator limit.
+     */
```
Sorry, giving. Thanks

eather009 committed:
Lint errors solved

eather009 committed:
Lint errors solved

eather009 committed:
Fixed: Closing parenthesis of a multi-line function call must be on a line by itself

eather009 committed:
Fixed: Closing parenthesis of a multi-line function call must be on a line by itself

eather009 committed:
remove white space and multiline

eather009 committed:
remove white space and multiline

eather009 committed:
remove training space

eather009 committed:
Merge branch '3.x_paginator_helper'

eather009 committed:
Add Form helper

eather009 committed:
add space

dereuromark commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+        $out = $this->Form->create('', ['type' => 'get']);
+
+        if (empty($limit)) {
+            $limits = [10 => $this->Number->format(10), 25 => $this->Number->format(25), 50 => $this->Number->format(50), 100 => $this->Number->format(100)];
```
Shouldn't the default value (and selected) be the default 20 coming from the component?

eather009 replied:
Opps, sorry for mistaken variable name. Updated and pulled. Thank you

dereuromark commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+     *
+     * @return string html output.
+     */
+    public function showLimit($limits = [], $default = 0)
```
Why "show"? Maybe just `limit()` (like "numbers" and "last"/"prev") would also work?
Let's wait for more feedback here.

eather009 replied:
limit() is already used or confusing, may be some other name can be suggested please. I use showLimit because to mean showing limit options. Thanks for suggestion.

dereuromark replied:
Other ideas: `limitSelect()`, `limitControl()`. The latter could be in sync with the new input=>control naming change.

eather009 replied:
Thanks. I have changed the function name to limitSelect().

eather009 committed:
Mistaken variable name

eather009 committed:
showLimit function is renamed with limitSelect

dereuromark commented:
@eather009 You were a bit fast with the method rename.
I would have liked some more input here regarding a good name, e.g. my favorite would be `limitControl()`.

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+     */
+    public function limitSelect($limits = [], $default = 0)
+    {
+        $out = $this->Form->create('', ['type' => 'get']);
```
String is not a valid value for 1st argument for `FormHelper::create()` use `null` instead.

eather009 replied:
Thank you. Its updated.

eather009 replied:
Is this request change is accepted?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+        $out = $this->Form->create('', ['type' => 'get']);
+
+        if (empty($limits)) {
+            $limits = [10 => $this->Number->format(10), 25 => $this->Number->format(25), 50 => $this->Number->format(50), 100 => $this->Number->format(100)];
```
Why use `Number->format()` ?

eather009 replied:
To ensure internationalization. I am using bn_BD as default_locale, so I need the number to show as Bangla font.

eather009 replied:
Removed Number->format() as user has options to overide limites

eather009 replied:
Is this request change is accepted?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+        }
+
+        if (empty($default)) {
+            $default = $this->param('limit');
```
What if `$default` is not in `$limits` list?

eather009 replied:
Thank you for concern. I have given an update.

eather009 replied:
Is this request change is accepted?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+     *
+     * @return string html output.
+     */
+    public function limitSelect($limits = [], $default = 0)
```
Better to use `null` instead of `0` as default value for `$default`.

eather009 replied:
Thank you. Its updated

eather009 replied:
Is this request change is accepted?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+     * Dropdown array for Select limit
+     *
+     * @param array $limits This is array of options.
+     * @param array $default Default limit for option selecting. Default is Paginator limit.
```
Incorrect type, it's not an `array`.

eather009 replied:
Thank you. Its updated

eather009 replied:
Is this request change is accepted?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,31 @@ public function implementedEvents()
+            $default = $this->param('limit');
+        }
+
+        $out .= $this->Form->input('limit', ['type' => 'select', 'label' => __("View") . '&nbsp;', 'escape' => false, 'value' => $default, 'id' => 'select_limit', 'default' => $default, 'class' => 'form-control', 'options' => $limits, 'onChange' => 'this.form.submit()']);
```
All the attributes like `label`, `id`, `class` should be configurable by user. The bootstrap centric `form-control` is not a good default.

eather009 replied:
Thank you. I have updated it. I sent an options parameter which can be used for Form input.

eather009 replied:
Is this request change is accepted?

eather009 committed:
1. Set null for $default and Form->create
2. Check $default value and set if not in $limits
3. Sort $limits by Key
4. Dynamic attribute options for Select tag
5. Get default limit from perPage instead of limit param of Paginator

eather009 committed:
overide select tag attribute

dereuromark commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,38 @@ public function implementedEvents()
+
+        ksort($limits);
+
+        $out .= $this->Form->input('limit', ($options + ['type' => 'select', 'label' => __("View") . '&nbsp;', 'escape' => false, 'value' => $default, 'id' => 'select_limit', 'default' => $default, 'options' => $limits, 'onChange' => 'this.form.submit()']));
```
With `+ $options` you are forcing the label etc to not being able to changed anymore.
Was that the intention?

eather009 replied:
Oh, its been changed by my last update.

eather009 committed:
stickler-ci: Error fixed
Fixed: Doc comment for parameter "$options" missing
Fixed: Expected "int" but found "integer" for parameter type
Fixed: Missing parameter name

eather009 committed:
Check Formhelper is loaded or not. If not, return empty string

eather009 committed:
remove blank line

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1133,6 +1133,11 @@ public function implementedEvents()
      */
     public function limitSelect($limits = [], $default = null, $options = [])
     {
+        if (!in_array('Form', $this->helpers)) {
```
When would this case occur? You have included `Form` in `$helpers` property declaration.

eather009 replied:
Yes, but this is just for confirmation. To prevent from error if anytime occurs

eather009 replied:
I have removed this, as its unnecessary.

eather009 replied:
Is the request change done?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1133,6 +1133,11 @@ public function implementedEvents()
     {
+        if (!in_array('Form', $this->helpers)) {
+
+            return "";
```
Please use single quotes for consistency.

eather009 replied:
Thank you. Updated.

eather009 replied:
Is this request change done?

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,42 @@ public function implementedEvents()
+
+        ksort($limits);
+
+        $out .= $this->Form->input('limit', ($options + ['type' => 'select', 'label' => __("View") . '&nbsp;', 'escape' => false, 'value' => $default, 'id' => 'select_limit', 'default' => $default, 'options' => $limits, 'onChange' => 'this.form.submit()']));
```
It's not a good idea to set `id` attribute by default. It's pretty common to show the pagination links and limit selectors above and below the items listing in which case you would end up with duplicate `id`s.

eather009 replied:
Thank you. It has no use actually. I used it in my project and thus forgot to remove. Thank you. Its removed.

eather009 committed:
removed id

eather009 commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1121,4 +1121,38 @@ public function implementedEvents()
         return [];
     }
+
+    /**
```
Given

eather009 committed:
single quotes for consistenc

eather009 commented:
I am getting error on "The Travis CI Build" [https://travis-ci.org/cakephp/cakephp/jobs/201782883](url) that 

> 1188   Access to an undefined property           
         Cake\View\Helper\PaginatorHelper::$Form.  
  1204   Access to an undefined property           
         Cake\View\Helper\PaginatorHelper::$Form.  
  1206   Access to an undefined property           
         Cake\View\Helper\PaginatorHelper::$Form.  

I am confused, why is that? As I have called Form in Helper. Any suggestion?

eather009 committed:
Add Form Helper in Property

eather009 committed:
update the latest formation of property define. As previous was conflicted because of missing the update

eather009 committed:
Merge branch 'master' into master

eather009 committed:
Test case for Paginator::limitSelect()

eather009 committed:
update tag name

eather009 committed:
method missing fix

eather009 committed:
update test case

eather009 committed:
removed id for select test case

eather009 committed:
update Test Case, run without error in local server

eather009 committed:
remove extra code and Number format

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,38 @@ public function implementedEvents()
+
+        ksort($limits);
+
+        $out .= $this->Form->input('limit', ($options + ['type' => 'select', 'label' => __("View") . '&nbsp;', 'escape' => false, 'value' => $default, 'default' => $default, 'options' => $limits, 'onChange' => 'this.form.submit()']));
```
I don't think `'escape' => false` is necessary, also `'label' => __('View')` would probably be a better.

eather009 replied:
Thank you. Done.

ADmad commented on tests/TestCase/View/Helper/PaginatorHelperTest.php:
```diff
@@ -2546,4 +2546,57 @@ public function testMeta($page, $prevPage, $nextPage, $pageCount, $options, $exp
     }
+
+    /**
+     * @param array $limits
```
Params and return type are incorrect.

eather009 replied:
Thank you. Corrected.

ADmad commented:
I too prefer the name `limitControl()`.

eather009 committed:
remove 'escape' => false and update 'label' => __('View')
Also update testcase

eather009 committed:
update test case

eather009 committed:
update test case

eather009 committed:
make multiline array

eather009 committed:
check for array

eather009 committed:
Add 20 instead of 10 as this is default limit for Paginator

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,43 @@ public function implementedEvents()
+    {
+        $out = $this->Form->create(null, ['type' => 'get']);
+
+        if (empty($limits) || !is_array($limits)) {
```
The `!is_array()` check can be avoided it you add `array` typehint to the `$limits` argument.

eather009 replied:
Im getting Error in codecov/patch and codecov/project, thus gave this conditions.

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,43 @@ public function implementedEvents()
+
+        ksort($limits);
+
+        $out .= $this->Form->input('limit', ($options + ['type' => 'select', 'label' => __("View"), 'value' => $default, 'default' => $default, 'options' => $limits, 'onChange' => 'this.form.submit()']));
```
Setting `default` becomes redundant if you set `value`.

eather009 replied:
Thank you, Its removed

dereuromark replied:
You should not set value IMO, but only default. Otherwise you hardcode it and one cannot change it anymore at runtime.

dereuromark replied:
Looks like you removed the wrong one.

eather009 committed:
use control instead of input as this is deprecated in 3.4

dereuromark commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,38 @@ public function implementedEvents()
+    {
+        $out = $this->Form->create(null, ['type' => 'get']);
+
+        if (empty($default) || !is_integer($default)) {
```
Usually we use is_numeric as the value could also be stringish '1' etc (e.g. from form input), which would still be valid here though.

eather009 replied:
Thank you. Yes, I should also use is_numeric to allow stringish number. :)

dereuromark commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,38 @@ public function implementedEvents()
+     * Dropdown array for Select limit
+     *
+     * @param array $limits This is array of options.
+     * @param int $default Default limit for option selecting. Default is Paginator limit.
```
|null

eather009 replied:
Given. Thanks

eather009 committed:
set Param as null allowable.

eather009 committed:
Check by is_numeric to allow stringish number

eather009 committed:
added for appveyor

eather009 committed:
Wrong file selected

eather009 committed:
Update codecov.yml

markstory commented:
This seems like something that should be put into 3.5.0.

eather009 committed:
Function name is renamed as its prefered.

eather009 committed:
remove new line

eather009 committed:
params message updated

eather009 committed:
update appveyor.yml, new line added

eather009 committed:
reverted as cakephp have

eather009 committed:
-dummy commit, due to error in AppVeyor

ADmad commented:
- Please revert the changes to `.codecov.yml` and `appveyor.yml`
- Squash the commits into a single commit
- Re-target the changes to `3.next` branch.

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -1169,4 +1170,37 @@ public function implementedEvents()
+     *
+     * @return string html output.
+     */
+    public function limitControl($limits = [], $default = null, $options = [])
```
Add `array` typehint to `$limits` and `$options` params.

eather009 replied:
Done.

eather009 committed:
Revert "update appveyor.yml, new line added"

This reverts commit 174a8ec8ef352cf13215fa89fa7a7db4bbc275cc.

eather009 committed:
reverse appveyor.yml

eather009 committed:
array typehint to $limits and $options

dereuromark commented:
You dont have to close. You can Push to and Reuse the Same branch and PR.

eather009 commented:
Due to some issue, deleted the repository and fork again. Currently give update in 3.next branch.

eather009 commented:
Added to 3.next branch
https://github.com/cakephp/cakephp/pull/10301
Thank you
